### ISSUE TITLE
p:treetable is firing ajax event on row select with checkbox

### ISSUE THREAD

**Issue by vinitkishore** [primefaces/primefaces#3042]
# IMPORTANT 
- !!! If you open an issue, fill every item. Otherwise the issue might be closed as invalid. !!!

- Before you open an issue, test it with the current/newest version.
- Try to find an explanation to your problem by yourself, by simply debugging. This will help us to solve your issue 10x faster
- Clone this repository https://github.com/primefaces/primefaces-test.git in order to reproduce your problem, you'll have better chance to receive an answer and a solution.
- Feel free to provide a PR (Primefaces is an open-source project, any fixes or improvements are welcome.)

## 1) Environment
- PrimeFaces version: 6.1.9
- Does it work on the newest released PrimeFaces version? Version? No
- Does it work on the newest sources in GitHub? (Build by source -> https://github.com/primefaces/primefaces/wiki/Building-From-Source)
- Application server + version:
- Affected browsers:

## 2) Expected behavior
The ajax event should not fire if <p:ajax event="select" .... /> is not present within treetable.
...

## 3) Actual behavior
On row selection through checkbox, treetable is firing ajax event
..

## 4) Steps to reproduce
a. create a tree table with checkbox
b. From UI, select any row by clicking on the checkbox
c. ajax event is fired.
.. 

## 5) Sample XHTML

		<p:treeTable id="sample" value="#{selection.value}"
			var="val" selectionMode="checkbox"
			styleClass="xyz" scrollable="true" scrollWidth="1000"
			selection="#{selection.selectedValue}" 
			emptyMessage="No data">
			
			<p:columns var="column" headerText="#{column.header}"
				value="#{selection.selectedColumns}"/>
				
				<h:outputText
					value="#{val[column.property]}" />
					
			</p:columns>
		</p:treeTable>

## 6) Sample bean

..

**Comment by melloware** [comment]
I can't quite explain why but it looks like it was intentional to always fire an ajax request in treetable.js...

```javascript
    fireSelectNodeEvent: function(nodeKey) {
        if(this.isCheckboxSelection()) {
            var $this = this,
            options = {
                source: this.id,
                process: this.id
            };
            
            options.params = [
                {name: this.id + '_instantSelection', value: nodeKey}
            ];
            
            options.oncomplete = function(xhr, status, args) {
                if(args.descendantRowKeys && args.descendantRowKeys !== '') {
                    var rowKeys = args.descendantRowKeys.split(',');
                    for(var i = 0; i < rowKeys.length; i++) {
                        $this.addToSelection(rowKeys[i]);
                    }
                    $this.writeSelections();
                }
            }
            
            if(this.hasBehavior('select')) {
                var selectBehavior = this.cfg.behaviors['select'];
                selectBehavior.call(this, options);
            }
            else {
                PrimeFaces.ajax.AjaxRequest(options);
            }
        }
        else {
            if(this.hasBehavior('select')) {
                var selectBehavior = this.cfg.behaviors['select'],
                ext = {
                    params: [
                        {name: this.id + '_instantSelection', value: nodeKey}
                    ]
                };
                
                selectBehavior.call(this, ext);
            }
        }
    }
```

**Comment by jxmai** [comment]
https://github.com/primefaces/primefaces/commit/24c82e7af11761304fb56a2a86951fc5f1d2b6f6#diff-3676751f7229ae9ae602ba54c488578fR428

Might be a small mistake embedded in a big commit. Should we also check `if(this.hasBehavior('select'))` ?

**Comment by melloware** [comment]
@jxmai the one that has me scratching my head is this block here where it intentionally is sending data back even if the select event is not used like it still wants to do it for checkboxes.

```javascript
            if(this.hasBehavior('select')) {
                var selectBehavior = this.cfg.behaviors['select'];
                selectBehavior.call(this, options);
            }
            else {
                PrimeFaces.ajax.AjaxRequest(options);
            }
```

**Comment by jxmai** [comment]
Hmm. Maybe the trigger is before this function. As you can see, the function is called `fireSelectedNodeEvent` so we can expect it to fire ajax.

https://github.com/primefaces/primefaces/blob/24c82e7af11761304fb56a2a86951fc5f1d2b6f6/src/main/resources/META-INF/resources/primefaces/treetable/treetable.js#L304-L307


The code was taken from `toggleCheckboxNode`, and you can see checkbox can fire event no matter what.

**Comment by jxmai** [comment]
@melloware By the way, for some reason there is a semicolon right after the bracket. I don't recall anything like this in JS syntax.

https://github.com/primefaces/primefaces/blob/24c82e7af11761304fb56a2a86951fc5f1d2b6f6/src/main/resources/META-INF/resources/primefaces/treetable/treetable.js#L268-L270

**Comment by jxmai** [comment]
https://www.primefaces.org/showcase/ui/data/treetable/selection.xhtml

The checkbox selection fires ajax every time, and the single selection does not. The advantage of checkbox is to provide the ability to select all children of a node by a single click. I don't really see the purpose of that mandatory AJAX event to be honest other than slow down the UI / increase network traffic unnecessarily.

**Comment by melloware** [comment]
Agreed and that extra semicolon has to just be a mistake right?  I have never seen that either.

**Comment by jxmai** [comment]
Alright.   #3047

**Comment by tandraschko** [comment]
I dont think we can remove it, otherwise its impossible to transfer the selection to the server.

**Comment by jxmai** [comment]
@tandraschko Good catch. I see what you meant. OK, that makes sense. I will modify the PR to only clean up the `;` then

@melloware  We have to  fire AJAX to guarantee the selection function to work correctly

In short, the collpased fields cannot be sent to server by UI mapping.

without firing AJAX:

![image](https://user-images.githubusercontent.com/10467831/33740011-ca0ff02a-db64-11e7-837e-f213233fad5c.png)


the correct behavior should be 

![image](https://user-images.githubusercontent.com/10467831/33740054-e71910c0-db64-11e7-9904-e3dd5b20fdf4.png)
